src/Workspaces/Core/Portable/Formatting/Engine/TokenStream.Iterator.cs

```diff
@@ -10,11 +11,11 @@ internal partial class TokenStream
     {
         // gain of having hand written iterator seems about 50-100ms over auto generated one.
         // not sure whether it is worth it. but I already wrote it to test, so going to just keep it.
-        private class Iterator : IEnumerable<ValueTuple<int, SyntaxToken, SyntaxToken>>
+        private struct Iterator : IEnumerable<ValueTuple<int, SyntaxToken, SyntaxToken>>
         {
-            private readonly List<SyntaxToken> _tokensIncludingZeroWidth;
+            private readonly ImmutableArray<SyntaxToken> _tokensIncludingZeroWidth;
```

Comment:
i'm genuinely curious (but you don't need to change) if we would benefit from a pooled ArrayBuilder here so that we don't even need to allocate these over and over.